---
title: "Assignment 2 group 6"
author: "Adham Khaled, Juan Manuel Medina, Antonio Ortega, Isabella Skandorff & Andreas Vincent"
date: "May 15, 2017"
output: pdf_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(include = TRUE, echo = TRUE)
```


```{r, message = F}
library("ggplot2")
library("cowplot")
library("reshape2")
library("dplyr")
library("gridExtra")
library("VennDiagram")
theme_set(theme_bw())
```

# Part 1
## (a) What are the first five genomic nucleotides from the first exon of this transcript? 
AAAGG
\newline 

The DICER1 mRNA molecule should have the same sequence as the DNA genomic sequences in the sense strand (substituting Ts by Us). As AK002007 is transcribed on the reverse strand and the default genomic sequence presented by the browser is the antisense one, we have to reverse it. Therefore, the first five nucleotides of the exon in the AK002007 cDNA are AAAGG.

## (b) Look at the raw mRNA sequence of AK002007, from the database it actually comes from. What are the first five nucleotides?

When we check the raw mRNA sequence of AK002007, it can be seen that the first five nucleotides are GAAGC.

## (c) How do you explain the discrepancy (maximum 5 lines)?

AK002007 is possibly a truncated version of the DICER1 mRNA found in Genbank and refseq. The first 11 nt of  AK002007 are found in the last 11 nt of the previous exons in the other mrna sequences. The first 4 nt of these are postulated to be part of the 5' UTR of AK002007 by the aligner, confirming a misalignment. Otherwise, those 11 nt would be aligned as a separate exon in alignment with the other mrna sequences.

![](AK002007_1.png)
**Figure 1**: Screenshot of the start of AK002007. The first 7 letters are discarded by the aligner, while the following 4 are aligned to the end of the intron.

![](AK002007_2.png)
**Figure 2**: Screenshoot of the upstream exon. As we can see, the last letters of the exon match the letters aligned to the intron.


```{r, include=TRUE, echo = F}
knitr::include_graphics(path = c("focus.jpg"))
```

**Figure 3**: Shows the mRNA sequence of AK002007 and the general DICER1. These mRNA sequences are read from the right to the left, because they are transcribed from the minus strand. Remarkably, the final 4 letters of the upstream exon and the intron are the exact same GAAA

# Part 2
## (a)

Bedtools commands to preprocess data before R
```{r, include = T, eval = F, echo = T, engine = "bash"}
cat ERa_hg18.bed ERb_hg18.bed > full.bed
bedtools sort -i full.bed > full.sorted.bed
bedtools merge -i full.sorted.bed > merged.bed
# Calculates the percent coverage by chromosome
bedtools genomecov -i merged.bed -g hg18_chrom_sizes.txt > coverage.txt
bedtools genomecov -i ERa_hg18.bed -g hg18_chrom_sizes.txt > ERa.txt
bedtools genomecov -i ERb_hg18.bed -g hg18_chrom_sizes.txt > ERb.txt
```

R
```{r fig.height = 3, warning = F, messages = F}
df <- rbind(cbind(read.table(file = "coverage.txt"), protein = "Merged"),
            cbind(read.table(file = "ERa.txt"), protein = "ERa"),
            cbind(read.table(file = "ERb.txt"), protein = "ERb"))
colnames(df) <- c("chr", "bit", "start", "end", "fraction", "protein")
df$chr <- factor(as.character(df$chr),
                 levels = c(paste("chr", c(as.character(1:22), "X", "Y", "M"), sep = ""), "genome"))
sorted.index <- order(as.numeric(substring(df$chr, 4)))
df <- df[sorted.index,]
ggplot(data = df[(df$bit == 1 & df$chr != "genome"),],
       mapping = aes(x = chr, y = fraction, fill = protein)) + 
  geom_bar(stat = "identity", position = "dodge") +
  scale_y_continuous(limits = c(0, 6e-4)) +
   theme(legend.title = element_blank(),
         legend.position = c(0.75, 0.87),
         legend.key.size = unit(0.5, "cm"),
         legend.direction="horizontal")
```

**Figure 4.** _holi_

## (b)

Bedtools commands to preprocess data before R
```{r, eval = F, engine = "bash"}
# Calculates number of interval overlaps between ERa and ERb
# Reports multiple overlaps as a single occurrence
bedtools intersect -a ERa_hg18.bed -b ERb_hg18.bed -c > AtoBoverlap.bed
bedtools intersect -a ERb_hg18.bed -b ERa_hg18.bed -c > BtoAoverlap.bed
```

```{r fig.height=3, fig.width=3, echo=FALSE, include = 4, message = F, warning = F, fig.align="center"}
df2 <- read.table(file ="AtoBoverlap.bed")
df3 <- read.table(file ="BtoAoverlap.bed")
overlap <- sum(df3$V4)
draw.pairwise.venn(nrow(df2), nrow(df3),
                   cross.area = overlap,
                   category = c("ERa", "ERb"),
                   lty = rep("blank", 2),
                   fill = c("skyblue", "red"),
                   alpha = rep(0.5, 2))
```
## 8. Appendix

```{r, include = T}
sessionInfo()
```