---
title: "Assignment 2 group 6"
author: "Adham Khaled, Juan Manuel Medina, Antonio Ortega, Isabella Skandorff & Andreas Vincent"
date: "May 15, 2017"
output: pdf_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(include = FALSE, echo = FALSE)
```


```{r}
library("ggplot2")
library("cowplot")
library("reshape2")
library("dplyr")
library("VennDiagram")
theme_set(theme_bw())
```

# Part 1
## (a)What are the first five genomic nucleotides from the first exon of this transcript? 
AAAGG
\newline 

The DICER1 mRNA molecule should have the same sequence as the sense DNA genomic sequences (substituting Ts by Us). As AK002007 is transcribed on the reversed strand and the default genomic sequence presented by the browser is the antisense one, we have to reverse it. Therefore, the first five nucleotides of the exon in the AK002007 cDNA are AAAGG.

## (b) Look at the raw mRNA sequence of AK002007, from the database it actually comes from. What are the first five nucleotides?

When we check the raw mRNA sequence of AK002007, it can be seen that the first five nucleotides are GAAGC.

## (c) How do you explain the discrepancy (maximum 5 lines)?


### TODO Include capture from UCSCBrowser and Isabella's illustration showing the mess

**What is going on?**

The sequencing process generated a truncated version of the mature mRNA starting at one of the last exons of the DICER1 gene. This truncated molecule bears the last 11 nucleotides of exon 21/27 ("gaagcaaaaag") and continues to hold the nucleotides in exon 22/27. Nevertheless, the aligner has mismapped these first nucleotides. Instead, the first 7 letters have been ignored and the remaining 4 letters have been mapped to the end of the intronic region between exons 21 and 22.

**How could that happen?**

This could be due to the aligner penalizing opening the intron gap just to align the leading 11 letters. This is only allowed because unfortunately the last 4 letters of exon 21 are identical to the last 4 letters in the intron, therefore providing to the aligner with the freedom to choose where to put these 4 bases

```{r, include=TRUE}
knitr::include_graphics(path = c("AK002007_1.png", "AK002007_2.png"))
```


# Part 2
## (a)

Bedtools commands to preprocess data before R
```{r, include = T, eval = F, echo = T, engine = "bash"}
cat ERa_hg18.bed ERb_hg18.bed > full.bed
bedtools sort -i full.bed > full.sorted.bed
bedtools merge -i full.sorted.bed > merged.bed
# Calculates the percent coverage by chromosome
bedtools genomecov -i merged.bed -g hg18_chrom_sizes.txt > coverage.txt
```

R
```{r fig.height = 3, include = 8, echo = 1:2, warning = F, messages = F}
df <- read.table(file = "coverage.txt",
                col.names = c("chr", "bit", "start", "end", "fraction"))

# sort by chromosome
df$chr <- factor(as.character(df$chr),
                 levels = c(paste("chr", c(as.character(1:22), "X", "Y", "M"), sep = ""), "genome"))
sorted.index <- order(as.numeric(substring(df$chr, 4)))
df <- df[sorted.index,]
head(df)

# plot overlapping fraction of the chromosomes
# stat="idenity" to plot the values in fraction directly instead of a count/frequency
p1 <- ggplot(data = df[(df$bit == 1 & df$chr != "genome"),],
       mapping = aes(x = chr, y = fraction)) + 
  geom_bar(stat = "identity", fill = "orange") +
  scale_y_continuous(limits = c(0, 6e-4))

ggsave("fraction_barplot.png")
```
**Figure 1.** **A** _Barplot showing the distribution of the exon counts. Even though most of the genes contain less than 60 exons, as many as 150 may be found in some of them._ **B** _Detail for genes with max. 20 exons. The mode can be visualized at 3-5 exons per gene (max found at 4). The number of exons per gene decreases steadily beyond it._

```{r, include = T, eval = F, echo = T, engine = "bash"}
bedtools genomecov -i ERa_hg18.bed -g hg18_chrom_sizes.txt > ERa.txt
bedtools genomecov -i ERb_hg18.bed -g hg18_chrom_sizes.txt > ERb.txt
```


```{r fig.height = 3, include = FALSE, echo = FALSE, warning = F, messages = F}
df1 <- rbind(cbind(read.table(file = "ERa.txt"), protein = "ERa"),
             cbind(read.table(file = "ERb.txt"), protein = "ERb"))
colnames(df1) <- c("chr", "bit", "start", "end", "fraction", "protein")
df1$chr <- factor(as.character(df1$chr),
                 levels = c(paste("chr", c(as.character(1:22), "X", "Y", "M"), sep = ""), "genome"))
sorted.index <- order(as.numeric(substring(df1$chr, 4)))
df1 <- df1[sorted.index,]
p2 <- ggplot(data = df1[(df1$bit == 1 & df1$chr != "genome"),],
       mapping = aes(x = chr, y = fraction, fill = protein)) + 
  geom_bar(stat = "identity", position = "dodge") +
  scale_y_continuous(limits = c(0, 6e-4)) +
   theme(legend.title = element_blank(),
         legend.position = c(0.75, 0.93),
         legend.key.size = unit(0.5, "cm"),
         legend.direction="horizontal")
```

```{r include = T, echo = F}
library("cowplot")
plot_grid(p1, p2, rel_widths = c(1,1), nrow = 1, ncol = 2)
```

## (b)

Bedtools commands to preprocess data before R
```{r, include = T, eval = F, echo = T, engine = "bash"}
# Calculates number of interval overlaps between ERa and ERb
# Reports multiple overlaps as a single occurrence
bedtools intersect -a ERa_hg18.bed -b ERb_hg18.bed -c > AtoBoverlap.bed
bedtools intersect -a ERb_hg18.bed -b ERa_hg18.bed -c > BtoAoverlap.bed
```

```{r fig.height=3, fig.width=3, echo=FALSE, include = 4, message = F, warning = F, fig.align="center"}
df2 <- read.table(file ="AtoBoverlap.bed")
df3 <- read.table(file ="BtoAoverlap.bed")
overlap <- sum(df3$V4)
draw.pairwise.venn(nrow(df2), nrow(df3),
                   cross.area = overlap,
                   category = c("ERa", "ERb"),
                   lty = rep("blank", 2),
                   fill = c("skyblue", "red"),
                   alpha = rep(0.5, 2))
```
## 8. Appendix

```{r, include = T}
sessionInfo()
```



```{r code=readLines(knitr::purl("assignment2_group6.Rmd", documentation = 0)), eval = FALSE}
```